### Dislaimer
This notebook requires knowledge in:
* Python
* Neural Networks
* Pytorch Datasets and Modules
* Machine Learning Process Understanding

### Word Embeddings Example
The goal of this notebook is to have a hands-on experience of words embeddings.\
We will do the following:
* Load a set of Arabic text as trigrams
* Build a simple neural network
* Train the network to **predict masked words**
* Use the network weights as embeddings

#### Build Dataset

To train an embedding, we will use the dataset we used in the previous notebook

In the previous notebook we added the token `<UNKOWN>` to be a surregate for any unkown word\
that might be entered to the model during testing or deployment. For training by masking we will\
add two more tokens `<PAD>` and `<MASK>`.

Unlike in the next word prediction task, masked word prediction\
task take a sequence of tokens of a specific length `max_length`. Some squences will be shorter\
than the maximum length, hence the usage of `<PAD>` to fill in the remaining slots in the\
input sequence. Some of these tokens will be masked and the the model will have to predict\
them, hence the usage of `<MASK>`.

During the dataset initlization we will load the entire dataset in `alaraby1k.json` and then\
generate sequnces based on sentences in the data set. We will split the text by `[! ? ، .]`\
and consider each sentence a sequence. If the sequence is longer than `max_length` we wil discard it.\
We set `max_length` to 100.

In [7]:
from torch.utils.data import Dataset
from collections import defaultdict
from sklearn.model_selection import train_test_split
import json, re

class MyDataset(Dataset):

    def __init__(self, alaraby_filepath, max_length, is_train):
        self.raw_data = [article["text"] for article in json.load(open(alaraby_filepath, "r"))]
        self.train_raw_data, self.test_raw_data = train_test_split(self.raw_data, test_size=0.1, random_state=42)
        
        self.max_length = max_length
        self.SPECIAL_TOKENS_IDS = {"<UNKOWN>": 0, "<PAD>": 1, "<MASK>": 2}
        self.SPECIAL_IDS_TOKENS = {0: "<UNKOWN>", 1: "<PAD>", 2: "<MASK>"}

        self.train_sequences =  self.__generate_seq__(self.train_raw_data, self.max_length)
        self.test_sequences =  self.__generate_seq__(self.test_raw_data, self.max_length)
        self.vocab, self.id_to_word, self.word_to_id = self.__compute_vocab__(self.train_raw_data)


        self.is_train = is_train
    
    def __generate_seq__(self, texts, max_length):
        sentences = [re.split('\.|،|!|؟', text) for text in texts]
        sequences = [s for sequences in sentences for s in sequences]
        sequences = [s.split() for s in sequences]
        sequences = [s for s in sequences if len(s) <= max_length]
        return sequences
    
    def __compute_vocab__(self, texts):
        words = set()
        for text in texts:
            words.update(set(text.split()))
        words_list = list(self.SPECIAL_TOKENS_IDS.keys()) + list(words)
        id_to_word = defaultdict(lambda: "<UNKNOWN>", {idx: value for idx, value in enumerate(words_list)})
        word_to_id = defaultdict(lambda: 0, {value: idx for idx, value in enumerate(words_list)})
        return words, id_to_word, word_to_id

    def __len__(self):
        return len(self.train_sequences) if self.is_train else len(self.test_sequences)

    def __getitem__(self, idx):
        sequences = self.train_sequences if self.is_train else self.test_sequences
        sequence = sequences[idx]
        sequence = ["<PAD>"]*(self.max_length - len(sequence)) + sequence
        sequence = [ self.word_to_id[token] for token in sequence]
        return sequence
    
    def get_word_from_id(self, idx):
        return self.id_to_word[idx]
    
    def get_word_id(self, word):
        return self.word_to_id[word]
    
    def get_vocab_size(self):
        return len(self.vocab) 

In [8]:
max_length = 100
train_dataset = MyDataset("../Dataset/alaraby1k.json", max_length, is_train= True)
test_dataset = MyDataset("../Dataset/alaraby1k.json", max_length, is_train= False)

In [9]:
print(f"Id of unkown word: {train_dataset.get_word_id('<UNKNOWN>')}")
print(f"ًWord of unkown id: {train_dataset.get_word_from_id(-1)}")
print(f"Vocab size: {train_dataset.get_vocab_size()}")

item = 0
seq = train_dataset.__getitem__(item)
print(f"Item {item} in ids: {str(seq)}")
print(f"Item {item} in words: [{str([train_dataset.get_word_from_id(t) for t in seq])}")

item = 33
seq = train_dataset.__getitem__(item)
print(f"Item {item} in ids: {str(seq)}")
print(f"Item {item} in words: [{str([train_dataset.get_word_from_id(t) for t in seq])}")

item = 1200
seq = train_dataset.__getitem__(item)
print(f"Item {item} in ids: {str(seq)}")
print(f"Item {item} in words: [{str([train_dataset.get_word_from_id(t) for t in seq])}")

Id of unkown word: 0
ًWord of unkown id: <UNKNOWN>
Vocab size: 163101
Item 0 in ids: [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 71314, 116719, 106632, 157858, 50174, 31002, 9641, 49873, 154270, 136311, 48266, 16153, 17301, 106632, 95204, 52240, 21155, 118809, 44523, 101061, 76844, 145301, 95617, 153465, 30667, 106632, 69867, 23890, 78859, 150919, 69612, 24505, 69054, 18047, 374]
Item 0 in words: [['<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD>', '<PAD

#### Build Neural Network

The network task is framed as a straightforward classification task. Consider the `max_length`\
of the input sequence. Consider the `vocab_size`. The input is `max_length x vocab_size`.\
We are predicting masked words so the output has the same size as the input.\
We need to predict the same sequence arn't we. The logits size will be\
also `max_length x vocab_size` then. We will then compute the `cross_entropy` loss.

However, asking the model to predict the same output except for some masked words is reduntand.\
We will be computing loss over output tokens that we already know. To solve the problem \
we will do the following. We will compute the `cross_entropy` loss with a an altered verions\
of the input. Consider the following:

$$
The \ input: 
\begin{bmatrix}
    1 , 1 , 1 , 1 (word 1)\\ 
    1 , 1 , 1 , 1 (.) \\
    1 , 1 , 1 , 1 (.) \\
    . , . , . , . (.) \\
    1 , 1 , 1 , 1 (word 2) \\ 
\end{bmatrix}


$$


We will be using `nn.embedding` layer from pytroch. Internally the layer is a weight matrix of\
the following size `(embedding size, vocab size)`. The `embedding size` it the size of vecotr\
we want to represent the words with it is arbitrary chosen and optimized using hyperparameter\
optimization. The `vocab size` is the size of the vocab or the number of words we have.\
The matrix is updated and optimized using back propagation. We

Our vocab size is 163101 we will use an embedding of vocab size 165,000. And we will\
choose the embedding size to be `1024`. Which means each word will be represented using\
an a vector of size `1024`. Given the complex nature of languages in general \
an given the hard nature of the given task, one should use a large embedding size\
this gives the model more capacity to learn more and more about the semantics of the data.

Finally, be informed that `nn.embedding` is pretty similar to `nn.linear` layer.\
With a main difference that embedding takes words indices in the vocabulary\
while the linear layer takes one-hot encoded vectors.

In [5]:
import torch
import torch.nn as nn

class WordPredictor(nn.Module):
    def __init__(self, vocab_size, embedding_dim):
        super(WordPredictor, self).__init__()
        self.embedding = nn.Embedding(vocab_size, embedding_dim)
        self.linear = nn.Linear(embedding_dim*2, vocab_size)
    
    def forward(self, x1, x2):
        embedded1 = self.embedding(x1)
        embedded2 = self.embedding(x2)
        concatenated = torch.cat((embedded1, embedded2), dim=1)
        output = self.linear(concatenated)
        return output

    def embed(self, x):
        return self.embedding(x)

    def word(self, x):
      distance = torch.norm(self.embedding.weight.data - x, dim=1)
      nearest = torch.argmin(distance)
      return nearest

#### Train Embedding

In this code chunk we simply train the network given a number of hyperparameter.

In [ ]:
from torch.utils.data import DataLoader
from torch.optim import SGD
import os

device = torch.device("cuda:0" ) if torch.cuda.is_available() else torch.device("cpu" )

batch_size= 1500
train_dataloader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
test_dataloader = DataLoader(test_dataset, batch_size=batch_size, shuffle=True)

vocab_size = 165_000
embbeding_dim = 1024
model = WordPredictor(vocab_size, embbeding_dim)
model.to(device)

criterion = nn.CrossEntropyLoss()
optimizer = SGD(model.parameters(), lr=0.01)

chkpnt_path = f"word_predictor_{vocab_size}_{embbeding_dim}.chk"
if os.path.exists(chkpnt_path):
  model.load_state_dict(torch.load(chkpnt_path, map_location=device))


progress_path = "/word_predictor_progress.json"
if os.path.exists(progress_path):
  progress = json.load(open(progress_path, "r"))
else:
  progress = {"chkpnt" : 0, "progress" : []}

# Training loop
epochs = 100
for epoch in (progress["chkpnt"], epochs):
  for i, batch in enumerate(train_dataloader):
    optimizer.zero_grad()
    x1, x2, target = batch
    x1 = x1.to(device)
    x2 = x2.to(device)
    target = target.to(device)
    output = model(x1, x2)
    loss = criterion(output, target)
    loss.backward()
    optimizer.step()
    if(i % 10==0):
      print(f"Epoch {epoch } Batch {i}, Loss: {loss.item()}")
  progress["progress"].append({"epoch": epoch, "loss": loss.item()})
  progress["chkpnt"] = epoch   
  json.dump(progress, open(progress_path, "w"))
  torch.save(model.state_dict(), chkpnt_path)